src/attr.rs

```diff
@@ -0,0 +1,49 @@
+use std::fmt;
+use std::borrow::Cow;
+
+use escape::Escaped;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Attribute<'a> {
+    pub name: Cow<'a, str>,
+    pub value: Cow<'a, str>,
+}
+
+#[derive(Clone, Debug)]
+pub struct AttributeSet<'a> {
+    pub attrs: Cow<'a, [Attribute<'a>]>,
+}
+
+impl<'a> AttributeSet<'a> {
+    pub fn empty() -> AttributeSet<'a> {
+        AttributeSet::new(Cow::Borrowed(&[]))
+    }
+
+    pub fn new(attrs: Cow<'a, [Attribute<'a>]>) -> AttributeSet<'a> {
+        AttributeSet { attrs: attrs }
+    }
+}
+
+impl<'a, T> From<T> for AttributeSet<'a>
+    where T: Into<Cow<'a, [Attribute<'a>]>>
+{
+    fn from(attrs: T) -> AttributeSet<'a> {
+        AttributeSet::new(attrs.into())
+    }
+}
```

Comment:
I think it needs to be exported since it's used in another macro (unless I'm remembering the macro hygiene/visibility rules wrong, which I could certainly be doing).

Also could do something like

    let attrs = attrs!(class = "foo");
    let events = vec![start_tag!("h1").with_attrs(attrs), start_tag!("small").with_attrs(attrs)];

(Probably some better representative of the pattern, but you should get the idea)

I need to think more on the other questions, again it may be good to not have .with_attrs or anything similar until there's some actual use case for it. I'm hoping to get time to have a go at something like maud this weekend and see how the usability is.